File: examples/03_lifetimes/main.cpp
// ============================================
// 03 — Lifetimes & Dangling Pointers (C++)
// ============================================
// Kompilieren: g++ -std=c++17 main.cpp -o main && ./main
//
// C++ hat KEIN Lifetime-System.
// Der Entwickler ist selbst verantwortlich — der Compiler hilft kaum.

#include <iostream>
#include <string>
#include <string_view>

// --- Beispiel 1: Welche Lebensdauer hat der Rückgabewert? ---
// C++: Compiler weiss es nicht — du musst es wissen.
const std::string& longer(const std::string& a, const std::string& b) {
    return a.size() >= b.size() ? a : b;
    // Lebensdauer hängt davon ab, welcher Parameter zurückgegeben wird.
    // Kein Compile-Fehler, wenn Aufrufer es falsch verwendet!
}

// --- Beispiel 2: Dangling Reference — klassischer Bug ---
const std::string& get_longer_broken() {
    std::string temp = "short";
    std::string s1   = "very long string";
    return longer(s1, temp);
    // s1 und temp werden zerstört — Dangling Reference!
    // Kein Compile-Fehler.
}

// --- Beispiel 3: string_view — ähnlich wie Rust &str ---
// string_view hat dasselbe Problem: Lifetime manuell verwalten
std::string_view first_word(const std::string& s) {
    size_t pos = s.find(' ');
    return std::string_view(s).substr(0, pos);
}

void string_view_dangling() {
    std::string_view word;
    {
        std::string sentence = "hello world";
        word = first_word(sentence);
        std::cout << "Innerhalb: " << word << std::endl;
    } // sentence wird zerstört
    // std::cout << word; // 💥 Dangling string_view — UB!
    // In Rust: ❌ COMPILE ERROR — Lifetime-Checker verhindert das
}

// --- Beispiel 4: Struct mit Pointer — Lifetime-Verantwortung beim Dev ---
struct Excerpt {
    const char* part; // roher Pointer — wer besitzt das?
                      // In Rust: struct Excerpt<'a> { part: &'a str }
                      // — Compiler erzwingt korrekte Lifetime
};

int main() {
    std::string s1 = "long string is long";
    std::string s2 = "xyz";

    const std::string& result = longer(s1, s2);
    std::cout << "Länger: " << result << std::endl;

    // Gefährlicher Aufruf — kein Compile-Fehler:
    // const std::string& dangling = get_longer_broken();
    // std::cout << dangling; // 💥

    string_view_dangling();
    return 0;
}

File: examples/03_lifetimes/main.rs
// ============================================
// 03 — Lifetimes
// ============================================
//
// Jede Referenz hat eine Lifetime — wie lange sie gültig ist.
// Meistens inferiert der Compiler sie automatisch (Lifetime Elision).
// Explizite Annotationen nötig wenn der Compiler es nicht selbst herleiten kann.
//
// WICHTIG: Lifetimes sind reine Compiler-Annotationen — kein Runtime-Overhead!

fn main() {
    // --------------------------------------------------
    // 1. Dangling Reference — Compiler fängt es ab
    // --------------------------------------------------
    let r;
    {
        let x = String::from("hello"); // x lebt nur im Block
        r = &x;
        println!("r = {}", r); // ✅ hier noch OK
    } // x wird hier gedroppt
    // println!("{}", r);              // ❌ COMPILE ERROR: x lebt nicht lang genug

    // --------------------------------------------------
    // 2. longer() — explizite Lifetime nötig
    // --------------------------------------------------
    // Der Compiler muss wissen: wie lange lebt der Rückgabewert?
    let s1 = String::from("langer string");
    let ergebnis;
    {
        let s2 = String::from("kurz");
        ergebnis = longer(s1.as_str(), s2.as_str());
        println!("Längerer: '{}'", ergebnis); // ✅
    }
    // println!("{}", ergebnis); // ❌ COMPILE ERROR: s2 lebt nicht mehr

    // --------------------------------------------------
    // 3. Lifetime Elision — Compiler inferiert automatisch
    // --------------------------------------------------
    let s = String::from("hello world");
    let wort = erstes_wort(&s); // Lifetime wird automatisch inferiert
    println!("Erstes Wort: '{}'", wort); // ✅
}

// 'a bedeutet: Rückgabewert lebt so lange wie das KÜRZERE der beiden Parameter
fn longer<'a>(x: &'a str, y: &'a str) -> &'a str {
    if x.len() > y.len() {
        x
    } else {
        y
    }
}

// Hier inferiert der Compiler die Lifetime automatisch:
// fn erstes_wort<'a>(s: &'a str) -> &'a str { ... }
fn erstes_wort(s: &str) -> &str {
    match s.find(' ') {
        Some(i) => &s[..i],
        None => s,
    }
}
